From 14d5ee7b48491ccc4e62a648474dcf24dad9e568 Mon Sep 17 00:00:00 2001
From: Fabian Vogt <fvogt@suse.de>
Date: Mon, 16 Dec 2024 19:08:13 +0100
Subject: [PATCH] tpm2-util: Also retry unsealing after policy_pcr returns
 PCR_CHANGED

It's not just Esys_Unseal that may fail due to PCR changes during the
session, but also Esys_PolicyPCR. Perform a retry in that case as well.

Fixes #35490

(cherry picked from commit e61032bf47e6a7e572643a0060c6dd610635c854)

[fvogt: rebase on top of v256]
[fvogt: fixes boo#1233752]
[fvogt: fixes bsc#1234313]
---
 src/shared/tpm2-util.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/src/shared/tpm2-util.c b/src/shared/tpm2-util.c
index 495789024b..6ce77f9a4f 100644
--- a/src/shared/tpm2-util.c
+++ b/src/shared/tpm2-util.c
@@ -3979,6 +3979,9 @@ int tpm2_policy_pcr(
                         ESYS_TR_NONE,
                         NULL,
                         pcr_selection);
+        if (rc == TPM2_RC_PCR_CHANGED)
+                return log_debug_errno(SYNTHETIC_ERRNO(EUCLEAN),
+                                       "Failed to add PCR policy to TPM: %s", sym_Tss2_RC_Decode(rc));
         if (rc != TSS2_RC_SUCCESS)
                 return log_debug_errno(SYNTHETIC_ERRNO(ENOTRECOVERABLE),
                                        "Failed to add PCR policy to TPM: %s", sym_Tss2_RC_Decode(rc));
@@ -5744,6 +5747,8 @@ int tpm2_unseal(Tpm2Context *c,
                                 !!pin,
                                 pcrlock_policy,
                                 &policy_digest);
+                if (r == -EUCLEAN && i > 0)
+                        goto retry_after_pcr_changed;
                 if (r < 0)
                         return r;
 
@@ -5783,6 +5788,8 @@ int tpm2_unseal(Tpm2Context *c,
                 if (rc != TPM2_RC_PCR_CHANGED || i == 0)
                         return log_debug_errno(SYNTHETIC_ERRNO(ENOTRECOVERABLE),
                                                "Failed to unseal HMAC key in TPM: %s", sym_Tss2_RC_Decode(rc));
+
+retry_after_pcr_changed:
                 log_debug("A PCR value changed during the TPM2 policy session, restarting HMAC key unsealing (%u tries left).", i);
         }
 
-- 
2.47.0

